### ISSUE TITLE
coveralls crashes when there is GITHUB_{ACTIONS,TOKEN} but no GITHUB_REF in the environment (e.g. with tox)

### ISSUE THREAD

**Issue by hartwork** [lemurheavy/coveralls-public#1607]
This is…
- with `GITHUB_ACTIONS=true` and  `GITHUB_TOKEN` set but
- _without_ variable `GITHUB_REF`:

```console
# coveralls
Submitting coverage to coveralls.io...
Traceback (most recent call last):
  File "[..]/.tox/py38-coverage/bin/coveralls", line 8, in <module>
    sys.exit(main())
  File "[..]/.tox/py38-coverage/lib/python3.8/site-packages/coveralls/cli.py", line 95, in main
    result = coverallz.wear()
  File "[..]/.tox/py38-coverage/lib/python3.8/site-packages/coveralls/api.py", line 254, in wear
    json_string = self.create_report()
  File "[..]/.tox/py38-coverage/lib/python3.8/site-packages/coveralls/api.py", line 332, in create_report
    data = self.create_data()
  File "[..]/.tox/py38-coverage/lib/python3.8/site-packages/coveralls/api.py", line 387, in create_data
    self._data.update(git_info())
  File "[..]/.tox/py38-coverage/lib/python3.8/site-packages/coveralls/git.py", line 80, in git_info
    branch = git_branch()
  File "[..]/.tox/py38-coverage/lib/python3.8/site-packages/coveralls/git.py", line 36, in git_branch
    github_ref.startswith('refs/heads/')
AttributeError: 'NoneType' object has no attribute 'startswith'
```

Seen live today at https://github.com/asyncee/django-easy-select2/runs/4619646186?check_suite_focus=true#step:5:254 .

For a workaround in context of tox inside GitHub Actions:
1. [Add `passenv`](https://github.com/TheKevJames/coveralls-python/blob/master/docs/usage/tox.rst#github-actions) to the related environment in `tox.ini`:
   ```
   passenv = GITHUB_*
   ```
2. Invoke with `--service=github` to workaround issue #1435

**Comment by afinetooth** [comment]
@hartwork thank you. Can I ask for a clarification?:

Are you using one of the [community-created python integrations](https://docs.coveralls.io/python) for coveralls, or are you creating your own integration to be used via Github Actions? I'm not seeing any references in [your source code](https://github.com/asyncee/django-easy-select2), so I'm curious how to best contextualize this. TIA.

**Comment by hartwork** [comment]
Hi @afinetooth , happy to clarify. The integration is hand-made and part of a pull request — https://github.com/asyncee/django-easy-select2/pull/96/files — that has not been merged to `master`, yet. That's why you didn't see it. While building that manually from the ground up, I hit the issue above.

**Comment by afinetooth** [comment]
Ok, @hartwork, got it. Thanks.

One thing might be worth mentioning if you want to continue building upon your integration:

All integrations except one use the Coveralls repo token (env var: `COVERALLS_REPO_TOKEN`) to identify a repo to the [Coveralls API](https://docs.coveralls.io/api-reference) (in POSTs to the `/jobs` endpoint). That one integration is the [Coveralls Github Action](https://github.com/marketplace/actions/coveralls-github-action), which uses the Github secret token (`${{ secrets.GITHUB_TOKEN }}`) to identify the repo.

Given that you are creating this integration to use Github Actions, but may want to extend it to other CI services, it might be worth changing the value you're sending for `service_name`. The reason is that the value `service_name: github` is "reserved" by the [Coveralls Github Action](https://github.com/marketplace/actions/coveralls-github-action), insofar as, by passing this value to the [Coveralls API](https://docs.coveralls.io/api-reference), it's letting the API know the POST is coming via the [Coveralls Github Action](https://github.com/marketplace/actions/coveralls-github-action), and the API executes some logic specific to what is expected from that integration.

The first such example being that, when the API receives `service_name: github` it looks for the `GITHUB_TOKEN` rather than the Coveralls repo token, to identify your repo.

A more generic implementation that would play with all CI services would be to change the `service_name` value to anything besides `github`, such as `github-actions`, and then, for the `repo_token` parameter, provide the value for the Coveralls repo token (found on your Coveralls repo page), and/or pass it in the env var `COVERALLS_REPO_TOKEN`.

Let me know if I can clarify that further or if you have any further questions.

**Comment by hartwork** [comment]
Hi @afinetooth,

thanks for your explanation. I'm only *starting* to understand why you suggest avoiding `--service=github`. (I have only settled on `--service=github` because it made the HTTP status 422 error go way, as you know from #1435). I would try trading `--service=github` by specifying `COVERALLS_REPO_TOKEN` instead but I don't have access to Coveralls project https://coveralls.io/github/asyncee/django-easy-select2 myself, and someone with permissions (CC @asyncee) would need to add that value as a named GitHub secret to the GitHub repository, it seems.

I looked at migrating to [mentioned GitHub Action `coverallsapp/github-action`](https://github.com/coverallsapp/github-action) now, but learned that it [only supports LCOV files](https://github.com/coverallsapp/github-action/issues/30) (which is C land) while I have a `.coverage` file (which is Python land and SQLite technically). So moving to that GitHub Action doesn't seem to be an option.

That additional logic for magic service name `github` that you mention: is it in or against our favor? Does it break anything for https://github.com/asyncee/django-easy-select2 that I am not aware of?  I am already wondering how to get [those comment posts to pull requests](https://github.com/asyncee/django-easy-select2/pull/83#issuecomment-858843649) back that we seem to have lost somehow, if possible.  Is that related?  Any ideas?

Best, Sebastian

**Comment by asyncee** [comment]
@hartwork i can give your access to coveralls of you need to tune django-easy-select2 configuration.

**Comment by hartwork** [comment]
@asyncee it will allow me to experiment with the above. Thank you!

**Comment by asyncee** [comment]
@hartwork unfortunately i found no way to give you access to coveralls via it's UI. So I added project's token `COVERALLS_REPO_TOKEN`  to repository secrets.

**Comment by hartwork** [comment]
@asyncee thanks! Looks like the transition worked, see https://github.com/asyncee/django-easy-select2/pull/97 .

**Comment by afinetooth** [comment]
@hartwork (Cc: @asyncee), it looks like you've got this worked out, but let me answer some open questions for future reference:

@asyncee in the future, you can add a github user as a collaborator to your project, or to its org so it has access, and, at that point, the user will have access to the repo at Coveralls.io and be able to see the repo token. If you, additionally, give that user Admin access to the repo, they will have the ability to modify the repo's settings at coveralls.io, including changing its repo token. 

@hartwork to answer your questions / comments:

> [...] I don't have access to Coveralls project https://coveralls.io/github/asyncee/django-easy-select2 myself, and someone with permissions (CC @asyncee) would need to add that value as a named GitHub secret to the GitHub repository, it seems.

I understand, and given that constraint, you could have implemented your solution with `--service_name=github` (or `service=github` if that's how you're having your users declare it), it's use would just be limited to Github Actions for CI. On the other hand, you could get around that by also implementing a version with `COVERALLS_REPO_TOKEN` as the identifier, for use with other CI systems; you'd just have the extra coding and the extra effort of providing usage instructions for the two different contexts in your README.

If @asyncee had not supplied the repo token, those would have been your options.

> I looked at migrating to [mentioned GitHub Action `coverallsapp/github-action`](https://github.com/coverallsapp/github-action) now, but learned that it [only supports LCOV files](https://github.com/coverallsapp/github-action/issues/30) (which is C land) while I have a `.coverage` file (which is Python land and SQLite technically). So moving to that GitHub Action doesn't seem to be an option.

That is true, the [Coveralls Github Action](https://github.com/coverallsapp/github-action) currently requires coverage reports in LCOV format.

FWIW, the user from [this comment](https://github.com/coverallsapp/github-action/issues/30#issuecomment-894711005) reports success using [coveragepy-lcov](https://pypi.org/project/coveragepy-lcov/) with the [Coveralls Github Action](https://github.com/coverallsapp/github-action).

> That additional logic for magic service name `github` that you mention: is it in or against our favor? Does it break anything for https://github.com/asyncee/django-easy-select2 that I am not aware of?  I am already wondering how to get [those comment posts to pull requests](https://github.com/asyncee/django-easy-select2/pull/83#issuecomment-858843649) back that we seem to have lost somehow, if possible.  Is that related?  Any ideas?

It's in your favor / doesn't break anything, **_if you're only using Github Actions for CI_**.

The rest of the "magic" is small and mostly limited to the API being able to render a correct link back to your Github Actions workflow next to the job on your build page. 

Let me check out the status of your **pull request comments** to see if there's an issue I can help resolve. BRB here.

**Update**:
I'm seeing that PR Comments (with valid IDs) were created for your last three (2) pull_request builds:

Order: DATE DESC:
1. **Build 1628080224**: https://coveralls.io/builds/45251883  (**PR**: 98 - **Comment ID**: 1001714095)
2. **Build 1624475090**: https://coveralls.io/builds/45238604  (**PR**: 97 - **Comment ID**: 1001207817)

So it looks like PR Comments are working.

There was no comment sent for **Build 1616664368** (**PR 96**):
https://coveralls.io/builds/45218685

But that's because we can't find a _base build_ for that _pull_request build_. Which is why the top of your coverage table reads: `FIRST BUILD ON RESTORE-COVERALLS AT 100.0%`.

That just means we did not receive a coverage report for the commit on the branch that PR was meant to be merged into (likely **master**). 

But again, it looks like PR Comments are working, the issue is just limited to that build. If I can help you fix that, lmk.

**Comment by hartwork** [comment]
@afinetooth coveragepy-lcov is an interesting pointer. Thanks for elaborating :pray:

**Comment by afinetooth** [comment]
@hartwork you bet. Just replied re: PR Comments above. :)

**Comment by hartwork** [comment]
@afinetooth we have moved away from `--service=github` (and  `GITHUB_TOKEN`) to `COVERALLS_REPO_TOKEN` as suggested [in the mean time](https://github.com/asyncee/django-easy-select2/pull/97/files) and now today [we're getting](https://github.com/asyncee/django-easy-select2/runs/5087340738?check_suite_focus=true) error `coveralls.exception.CoverallsException: Running on Github Actions but GITHUB_TOKEN is not set. Add "env: GITHUB_TOKEN: ${{ secrets.GITHUB_TOKEN }}" to your step config`. That's not cool :smiley: Why?! :smiley: 

CC @asyncee

**Comment by afinetooth** [comment]
Hi @hartwork. 

From [the PR you shared](https://github.com/asyncee/django-easy-select2/pull/97/files), I think the issue might be that you're not declaring any value for `--service` and something along the way is choosing `github` as the value, such that `github` is acting as the default. 

That would explain why the integration complains, "Running on Github Actions but GITHUB_TOKEN is not set."

I'm hoping you can just pass a different explicit value for `--service` and it will work. 

In other words, change [this line](https://github.com/asyncee/django-easy-select2/pull/97/files#diff-ef2cef9f88b4fe09ca3082140e67f5ad34fb65fb6e228f119d3812261ae51449L34):

```
bash -c 'set -x; [[ $\{GITHUB_ACTIONS:-false\} = false ]] || coveralls --service=github'
```

To:

```
bash -c 'set -x; [[ $\{GITHUB_ACTIONS:-false\} = false ]] || coveralls --service=github-actions'
```

Let me know if that doesn't fix it.

**Comment by hartwork** [comment]
Hi @afinetooth thanks for your reply! I had a look at the [causing code in coveralls-python](https://github.com/TheKevJames/coveralls-python/blob/c35bf51671a9608b6b5610e61a1510bf1924882c/coveralls/api.py#L55-L59) now. Because it's detection centers around variable `GITHUB_ACTIONS`, I have decided to [take that variable away from it now](https://github.com/asyncee/django-easy-select2/pull/101), and that seems to work well. As a result, it [should send `"coveralls-python"` for a service name](https://github.com/TheKevJames/coveralls-python/blob/c35bf51671a9608b6b5610e61a1510bf1924882c/coveralls/api.py#L149) now, and it did stop complaining. So the symptom is fixed.

What I don't get is how that code that was introduced in [this coveralls-pyrthon commit](https://github.com/TheKevJames/coveralls-python/commit/f4faa92dd8c9e9a523a6e778f36b13496ad79760) for release 3.0.0 in 2021-01 did *not* break PR https://github.com/asyncee/django-easy-select2/pull/97 back in 2021-12 but [broke PRs yesterday](https://github.com/asyncee/django-easy-select2/runs/5087340885?check_suite_focus=true) with supposedly the exact same code. Weird :thinking: Any ideas?

**Comment by afinetooth** [comment]
Hi, @hartwork. I would have to see the PR associated with [this build log](https://github.com/asyncee/django-easy-select2/runs/5087340885?check_suite_focus=true), but I suspect the difference relates to the order of configuration precedence the author changed in [that commit](https://github.com/TheKevJames/coveralls-python/commit/f4faa92dd8c9e9a523a6e778f36b13496ad79760).

That is to say that, declaring the service in an env var in your GHA config yaml is not the same as declaring it as a flag in your command line command. I don't know exactly what the order of precedence is for that integration, but I suspect that declaring it in your command line is the last word and that, if you don't declare it there, some other value---an env var you're passing, or a default in the code---is getting used instead.

**Comment by hartwork** [comment]
@afinetooth I think you're arguing for `--service` over variables. Let me think about that.

**Comment by afinetooth** [comment]
@hartwork I'm actually agnostic. I think you could also use an environment variable (suggested variable names: [`COVERALLS_SERVICE_NAME`](https://github.com/nickmerwin/node-coveralls/search?q=COVERALLS_SERVICE_NAME) or [`CI_NAME`](https://github.com/nickmerwin/node-coveralls/search?q=CI_NAME) per the conventions set in the [node-coveralls](https://github.com/nickmerwin/node-coveralls) integration), or a [`.coveralls.yml` file](https://docs.coveralls.io/supported-ci-services#configuration) in the the root of your project with the [`service_name` key](https://docs.coveralls.io/supported-ci-services#configuration), and either one would effectively override defaults in the code.

But `--service` is a good option if it simplifies your usage instructions (especially if your users don't need to mess with any other env vars or a new config file for any other reasons).
